### PR TITLE
Ingest service ID and advertise from dates from BT

### PR DESCRIPTION
**Pull request by tfm**
…MpxChannelDataIngester, added logic to save aliases and availableDates in BtChannelDataUpdater and BtMpxChannelDataIngester, Added additional fields to extract from Json in GsonBtMpxClient, Added Spring wiring and bean in BtChannelsModule class

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -228,6 +228,14 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
```
Comment (tfm) [inline pom.xml]: Why was this added?

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    private final SubstitutionTableNumberCodec codec = new SubstitutionTableNumberCodec();
+
+    public void addAliasesAndAvailableDateToChannel(PaginatedEntries paginatedEntries, ChannelResolver channelResolver) {
+
+        List<Entry> entries = paginatedEntries.getEntries();
+
+        Set<Long> channelIdsThatHaveAliasesAndAvailableDateAdded = Sets.newHashSet();
+
+        for(int i = 0; i < entries.size(); i++) {
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Can't we use a for (Entry entry : paginatedEntries.getEntries() ) loop here?

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            Entry currentEntry = entries.get(i);
+            String linearEpgChannelId = currentEntry.getLinearEpgChannelId();
+            DateTime advertiseAvailableDate = currentEntry.getAvailableDate();
+            long currentGuid = codec.decode(currentEntry.getGuid()).longValue();
+
+            Maybe<Channel> channelMaybe = channelResolver.fromId(currentGuid);
+
+            if(channelMaybe.hasValue()) {
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: What if there's no value?

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                Channel channel = channelMaybe.requireValue();
+                Set<Alias> aliases = channel.getAliases();
+
+                aliases.removeIf(alias -> {
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Don't use Java 8

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+                    if (alias.getNamespace().equals(linearEpgChannelId)) {
+                        return true;
+                    }
+                    return false;
+                });
+
+                //Using "namespace" as namespace for now.
+                channel.addAlias(new Alias("namespace", linearEpgChannelId));
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Pass in the alias namespace as a constructor arg

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+                //Turned off for now until it is added to atlas model and atlas deer. Did a git pull but couldn't get dias' changes probably because it is only on stage. (my branch is based off master)
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Remember whether we ingest the available from date needs to be configurable based on the environment. We should similarly only wipe it, if we're ingesting it.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+public class BtChannelDataUpdater {
+
+    private final SubstitutionTableNumberCodec codec = new SubstitutionTableNumberCodec();
+
+    public void addAliasesAndAvailableDateToChannel(PaginatedEntries paginatedEntries, ChannelResolver channelResolver) {
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: The resolver should be a field, not passed in to this method. Passing in here breaks encapsulation, since someone using this class shouldn't need to know which `ChannelResolver` a `BtChannelUpdater` needs.

File: pom.xml (hunk 1 above)
Comment (timothy22000) [inline pom.xml]: I needed this to write lambda expressions (probably auto-generated by IntelliJ when I used the little light bulb button to fix the problem). If I remove it, it would give me "Lambda expressions are not supported at this language level".

Comment (tfm) [inline pom.xml]: Unfortunately you can't just upgrade to Java 8, since it's not running on a Java 8 VM yet.

Comment (timothy22000) [inline pom.xml]: I removed this and tweak the Module settings in Project Structure and it seems to be ok now and won't give the error above.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 4 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: This explains the pom.xml comment above now.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 2 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Yes, we can. I was using it out of habit. :)

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 4 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Why are we not using Java 8? Is it because it would break the codebase?

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 5 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Not too sure what you mean. The only two things that I can think of relating to this comment is:

plproductmetadata$linearEpgChannelId: "urn:BT:linear:service:751764"

i) use alias.getNamespace() to replace "namespace"? but alias.getNamespace() is "urn:BT:linear:service:751764" which is being used as the value. Another is that alias.getNamespace() would be removed in the step before which would make alias.getNamespace() empty.

ii) Do you mean use plproductmetadata$linearEpgChannelId as namespace?

File: pom.xml (hunk 1 above)
Comment (timothy22000) [inline pom.xml]: Ignore the above. Saw your comment below about not using Java 8.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 3 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: My thought process was do nothing (Would not add that channel's channelId to the channelIdsThatHaveAliasesAndAvailableDate Set). Thus, it will not be filtered out by the Iterable.filter method below when removing aliases from channelIdsThatDidNotHaveAliasesAndAvailableDateSet.

How would you like me to handle the case with no value?

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        removeStaleAliasesAndAvailableDateFromChannel(channelIdsThatHaveAliasesAndAvailableDateAdded, channelResolver.all());
+    }
+
+    private void removeStaleAliasesAndAvailableDateFromChannel(Set<Long> channelIdsThatHaveAliasesAndAvailableDateAdded, Iterable<Channel> channels) {
+        Iterable<Channel> channelsThatHasStaleAliasesAndAvailableDates = Iterables.filter(channels,
+                                                                            shouldRemoveValuesFromChannel(channelIdsThatHaveAliasesAndAvailableDateAdded));
+
+        StreamSupport.stream(channelsThatHasStaleAliasesAndAvailableDates.spliterator(), false).forEach(channel -> {
```
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: I won't be able to use StreamSupport as well since it is Java 8. Will move to good ol loops or Guava's FluentIterable.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 6 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: This was the last task on the ticket and I wasn't sure how to do it yet. Wanted to ask you about it in the morning :) I need to investigate this a bit more.

Idea so far:
Each environment creates a BtAllChannelsChannelGroupUpdater with different publisher. (Publisher.BT_TV_CHANNELS_TEST2 and others) which seems to be set in ChannelGroup under the createOrQueryBtAllChannelsChannelGroup().

Approach: Get publisher from ChannelGroup to be able to tell which environment it is coming from.

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 3 above)
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Perhaps a log entry? Think about someone trying to debug a problem and what they'd like to see in logs when doing so.

Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: Ok!

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java (hunk 6 above)
Comment (timothy22000) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: I am currently only ingesting availableFrom date if it is only from prod. (latest commit)

File: src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.Set;
+
+import static com.google.common.base.Predicates.not;
+import static com.google.gson.internal.$Gson$Preconditions.checkNotNull;
+
+public class BtChannelDataUpdater {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BtChannelDataUpdater.class);
+
+    private final SubstitutionTableNumberCodec codec = new SubstitutionTableNumberCodec();
```
Comment (tfm) [inline src/main/java/org/atlasapi/remotesite/bt/channels/BtChannelDataUpdater.java]: this should be lowercaseonly codec

**Comment by MartinsIrbe** [comment]
LGTM
